### Heart Disease Prediction

Step 1. Import Libraries

In [198]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Step 2. Data collection and Processing

In [199]:
# loading the csv data to a Pandas DataFrame

heart_data = pd.read_csv(r"C:\Users\Maftuna\Downloads\heart.csv")

In [200]:
# print first 5 rows of the dataset

heart_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [201]:
# print last 5 rows of the dataset 

heart_data.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [202]:
# number of rows and  columns in the dataset 

heart_data.shape

(303, 14)

In [203]:
# getting some info about the data 

heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [204]:
# Setting full column names
heart_data.columns = [
    "Age",                         # Age of the person
    "Sex",                         # Gender
    "Chest Pain Type",             # Type of chest pain
    "Resting Blood Pressure",      # Blood pressure at rest
    "Serum Cholesterol",           # Cholesterol level in the blood
    "Fasting Blood Sugar",         # Blood sugar level when fasting
    "Resting Electrocardiographic Results",  # Resting ECG results
    "Maximum Heart Rate Achieved", # Maximum heart rate achieved
    "Exercise-Induced Angina",     # Angina induced by exercise
    "ST Depression Induced by Exercise",     # ST depression caused by exercise
    "Slope of the Peak Exercise ST Segment", # Slope of the ST segment
    "Number of Major Vessels Colored by Fluoroscopy", # Major vessels colored during fluoroscopy
    "Thalassemia",                 # Blood disorder (thalassemia)
    "Target (Presence of Heart Disease)" # Target variable (Heart disease presence)
]

# Verifying the updated column names
print(heart_data.columns)



Index(['Age', 'Sex', 'Chest Pain Type', 'Resting Blood Pressure',
       'Serum Cholesterol', 'Fasting Blood Sugar',
       'Resting Electrocardiographic Results', 'Maximum Heart Rate Achieved',
       'Exercise-Induced Angina', 'ST Depression Induced by Exercise',
       'Slope of the Peak Exercise ST Segment',
       'Number of Major Vessels Colored by Fluoroscopy', 'Thalassemia',
       'Target (Presence of Heart Disease)'],
      dtype='object')


In [205]:
# Rename columns to more descriptive name 

heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Age                                             303 non-null    int64  
 1   Sex                                             303 non-null    int64  
 2   Chest Pain Type                                 303 non-null    int64  
 3   Resting Blood Pressure                          303 non-null    int64  
 4   Serum Cholesterol                               303 non-null    int64  
 5   Fasting Blood Sugar                             303 non-null    int64  
 6   Resting Electrocardiographic Results            303 non-null    int64  
 7   Maximum Heart Rate Achieved                     303 non-null    int64  
 8   Exercise-Induced Angina                         303 non-null    int64  
 9   ST Depression Induced by Exercise          

In [206]:
# Checking for missing values 

heart_data.isnull().sum()

Age                                               0
Sex                                               0
Chest Pain Type                                   0
Resting Blood Pressure                            0
Serum Cholesterol                                 0
Fasting Blood Sugar                               0
Resting Electrocardiographic Results              0
Maximum Heart Rate Achieved                       0
Exercise-Induced Angina                           0
ST Depression Induced by Exercise                 0
Slope of the Peak Exercise ST Segment             0
Number of Major Vessels Colored by Fluoroscopy    0
Thalassemia                                       0
Target (Presence of Heart Disease)                0
dtype: int64

In [207]:
# Statistical measures about the data 

heart_data.describe()

Age         Sex  Chest Pain Type  Resting Blood Pressure  \
count  303.000000  303.000000       303.000000              303.000000   
mean    54.366337    0.683168         0.966997              131.623762   
std      9.082101    0.466011         1.032052               17.538143   
min     29.000000    0.000000         0.000000               94.000000   
25%     47.500000    0.000000         0.000000              120.000000   
50%     55.000000    1.000000         1.000000              130.000000   
75%     61.000000    1.000000         2.000000              140.000000   
max     77.000000    1.000000         3.000000              200.000000   

       Serum Cholesterol  Fasting Blood Sugar  \
count         303.000000           303.000000   
mean          246.264026             0.148515   
std            51.830751             0.356198   
min           126.000000             0.000000   
25%           211.000000             0.000000   
50%           240.000000             0.000000   
75%           274.500000             0.000000   
max           564.000000             1.000000   

       Resting Electrocardiographic Results  Maximum Heart Rate Achieved  \
count                            303.000000                   303.000000   
mean                               0.528053                   149.646865   
std                                0.525860                    22.905161   
min                                0.000000                    71.000000   
25%                                0.000000                   133.500000   
50%                                1.000000                   153.000000   
75%                                1.000000                   166.000000   
max                                2.000000                   202.000000   

       Exercise-Induced Angina  ST Depression Induced by Exercise  \
count               303.000000                         303.000000   
mean                  0.326733                           1.039604   
std                   0.469794                           1.161075   
min                   0.000000                           0.000000   
25%                   0.000000                           0.000000   
50%                   0.000000                           0.800000   
75%                   1.000000                           1.600000   
max                   1.000000                           6.200000   

       Slope of the Peak Exercise ST Segment  \
count                             303.000000   
mean                                1.399340   
std                                 0.616226   
min                                 0.000000   
25%                                 1.000000   
50%                                 1.000000   
75%                                 2.000000   
max                                 2.000000   

       Number of Major Vessels Colored by Fluoroscopy  Thalassemia  \
count                                      303.000000   303.000000   
mean                                         0.729373     2.313531   
std                                          1.022606     0.612277   
min                                          0.000000     0.000000   
25%                                          0.000000     2.000000   
50%                                          0.000000     2.000000   
75%                                          1.000000     3.000000   
max                                          4.000000     3.000000   

       Target (Presence of Heart Disease)  
count                          303.000000  
mean                             0.544554  
std                              0.498835  
min                              0.000000  
25%                              0.000000  
50%                              1.000000  
75%                              1.000000  
max                              1.000000

In [208]:
# Checking the distribution of Target Variables 

heart_data['Target (Presence of Heart Disease)'].value_counts()

Target (Presence of Heart Disease)
1    165
0    138
Name: count, dtype: int64

1 --> Defective Heart 

0 --> Healthy Heart

Step 3. Split the Features and Target 

In [209]:
X = heart_data.drop(columns='Target (Presence of Heart Disease)', axis=1)
Y = heart_data['Target (Presence of Heart Disease)']

In [210]:
print(X)

     Age  Sex  Chest Pain Type  Resting Blood Pressure  Serum Cholesterol  \
0     63    1                3                     145                233   
1     37    1                2                     130                250   
2     41    0                1                     130                204   
3     56    1                1                     120                236   
4     57    0                0                     120                354   
..   ...  ...              ...                     ...                ...   
298   57    0                0                     140                241   
299   45    1                3                     110                264   
300   68    1                0                     144                193   
301   57    1                0                     130                131   
302   57    0                1                     130                236   

     Fasting Blood Sugar  Resting Electrocardiographic Results  \
0        

In [211]:
print(Y)

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: Target (Presence of Heart Disease), Length: 303, dtype: int64


Step 4. Splitting the Data into Trainging data & Test Data

In [212]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=42)

In [213]:
print("X shape: ", X.shape)
print("X_train shape: ", X_train.shape)
print("X_test shape: ", X_test.shape)

X shape:  (303, 13)
X_train shape:  (242, 13)
X_test shape:  (61, 13)


In [214]:
from sklearn.preprocessing import StandardScaler




# Scalingni yaratish
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)  # Train uchun fit va transform
X_test_scaled = scaler.transform(X_test)       # Test uchun faqat transform


Step 5. Model Training

In [215]:
# Logictic Regression 

model = LogisticRegression()

In [216]:
model.fit(X_train, Y_train)

C:\Users\Maftuna\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model Evaluation

Accuracy Score 

In [217]:
# accurancy on training data

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [218]:
print("Accuracy on Training data : " , training_data_accuracy)

Accuracy on Training data :  0.8429752066115702


In [219]:
# accurancy on test data

X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [220]:
print("Accuracy on Test data : " , test_data_accuracy)

Accuracy on Test data :  0.8032786885245902



 ##### Whithout Scaling
##### 1. Accuracy on Training data :  0.8429752066115702
##### 2. Accuracy on Test data :  0.8032786885245902

Step 6. Building a Predictive System

In [221]:
input_data = (51,0,0,130,305,0,1,142,1,1.2,1,0,3)

# Change the input data to a array 
input_data_as_numpy_array = np.array(input_data)

# Reshape the numpy array as we are predicting for only on instance 
input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

prediction = model.predict(input_data_reshaped)
print(prediction)

if prediction[0] == 0:
    print("The Person does not have a Heart Disease")
else: 
    print("The Person has Heart Disease")

[0]
The Person does not have a Heart Disease


C:\Users\Maftuna\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [223]:
# Logistic Regression Coefficients
coefficients = model.coef_
print("Model Coefficients:", coefficients)

# Feature Importance for Decision Trees or Random Forests
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier()
rf_model.fit(X_train_scaled, Y_train)
importances = rf_model.feature_importances_
print("Feature Importances:", importances)


Model Coefficients: [[ 0.0127083  -1.16302116  0.88132356 -0.00763027 -0.00764697 -0.20184379
   0.51158657  0.03240623 -0.8171904  -0.49052807  0.42248244 -0.54033279
  -0.98663251]]
Feature Importances: [0.0828348  0.03067137 0.13039348 0.07531126 0.09034807 0.00971271
 0.02160696 0.10805151 0.07281984 0.10988868 0.06972225 0.08438775
 0.11425131]


In [226]:
from sklearn.neighbors import KNeighborsClassifier

# k-Nearest Neighbors modeli
knn_model = KNeighborsClassifier()
knn_model.fit(X_train_scaled, Y_train)
y_pred_knn = knn_model.predict(X_test_scaled)

accuracy_knn = accuracy_score(Y_test, y_pred_knn)
print("Accuracy with KNN (scaled):", accuracy_knn)


Accuracy with KNN (scaled): 0.8032786885245902
